File: scripts/scenarios/scenarios/39_list_management.py
"""Scenario 39: "The Organizer" — List Management

Luna creates and manages lists: curatelist and modlist, adds and
removes members, and verifies list operations.

Services: PDS, AppView
"""

# SPDX-FileCopyrightText: 2025-2026 Jack Valinsky
# SPDX-License-Identifier: Unlicense OR CC0-1.0

from __future__ import annotations

import sys
import time
from pathlib import Path

_project_root = str(Path(__file__).resolve().parent.parent.parent.parent)
if _project_root not in sys.path:
    sys.path.insert(0, _project_root)

from scripts.lib.atproto import (
    XrpcClient, get_character, PDS1, ScenarioResult, timed_call,
    SERVICE_URLS, create_account_or_login, now_iso,
)


def run() -> ScenarioResult:
    result = ScenarioResult("List Management")
    result.start()

    pds = XrpcClient(PDS1)
    appview = XrpcClient(SERVICE_URLS["appview"])
    luna = get_character("luna")
    bob = get_character("bob")

    timed_call(result, "PDS health check",
               lambda: pds.wait_for_healthy(timeout=30))
    if result.failed > 0:
        result.finish()
        return result

    # Create accounts
    luna_session = timed_call(
        result, "Create account for Luna",
        lambda: create_account_or_login(pds, luna),
        detail_fn=lambda s: f"did={s.get('did', '?')}",
    )
    bob_session = timed_call(
        result, "Create account for Bob",
        lambda: create_account_or_login(pds, bob),
        detail_fn=lambda s: f"did={s.get('did', '?')}",
    )

    if not luna_session or not bob_session:
        result.finish()
        return result

    luna.did = luna_session["did"]
    luna.access_jwt = luna_session["accessJwt"]
    bob.did = bob_session["did"]

    # Create a curate list
    list_rkey = "curate-list-" + str(int(time.time()))
    list_record = {
        "$type": "app.bsky.graph.list",
        "purpose": "app.bsky.graph.defs#curatelist",
        "name": "Luna's Favorites",
        "description": "Accounts Luna finds interesting",
        "createdAt": now_iso(),
    }

    list_ref = timed_call(
        result, "Create curate list",
        lambda: pds.repositories.create_record(
            luna.did, "app.bsky.graph.list", list_rkey,
            list_record, luna.access_jwt,
        ),
        detail_fn=lambda r: f"uri={r.get('uri', '?')}" if r else "failed",
    )

    if not list_ref:
        result.finish()
        return result

    list_uri = list_ref.get("uri", "")

    # Add Bob to the list
    item_rkey = "listitem-" + str(int(time.time()))
    listitem_record = {
        "$type": "app.bsky.graph.listitem",
        "list": list_uri,
        "subject": bob.did,
        "createdAt": now_iso(),
    }

    timed_call(
        result, "Add Bob to curate list",
        lambda: pds.repositories.create_record(
            luna.did, "app.bsky.graph.listitem", item_rkey,
            listitem_record, luna.access_jwt,
        ),
    )

    time.sleep(1)

    # Verify list appears in getLists
    timed_call(
        result, "Get lists for Luna",
        lambda: appview.graph.get_lists(
            {"actor": luna.did, "limit": 10}, luna.access_jwt,
        ),
        detail_fn=lambda r: f"count={len(r.get('lists', []))}" if r else "failed",
    )

    # Verify Bob appears in getList
    timed_call(
        result, "Get list items",
        lambda: appview.graph.get_list(
            {"list": list_uri, "limit": 10}, luna.access_jwt,
        ),
        detail_fn=lambda r: f"count={len(r.get('items', []))}" if r else "failed",
    )

    # Remove Bob from the list
    item_uri = list_uri.replace("app.bsky.graph.list", "app.bsky.graph.listitem") + "/" + item_rkey
    timed_call(
        result, "Remove Bob from list",
        lambda: pds.repositories.delete_record(
            luna.did, "app.bsky.graph.listitem", item_rkey,
            luna.access_jwt,
        ),
    )

    # Create a moderation list
    mod_list_rkey = "mod-list-" + str(int(time.time()))
    mod_list_record = {
        "$type": "app.bsky.graph.list",
        "purpose": "app.bsky.graph.defs#modlist",
        "name": "Luna's Block List",
        "description": "Accounts to moderate",
        "createdAt": now_iso(),
    }

    timed_call(
        result, "Create mod list",
        lambda: pds.repositories.create_record(
            luna.did, "app.bsky.graph.list", mod_list_rkey,
            mod_list_record, luna.access_jwt,
        ),
    )

    # Verify both lists accessible
    timed_call(
        result, "Verify both lists accessible",
        lambda: appview.graph.get_lists(
            {"actor": luna.did, "limit": 10}, luna.access_jwt,
        ),
        detail_fn=lambda r: f"count={len(r.get('lists', []))}" if r else "failed",
    )

    result.finish()
    return result

File: scripts/scenarios/scenarios/39_list_management.ts
/**
 * @module scenarios/39_list_management
 *
 * Scenario: Creates a curated list, adds a member, and removes it again.
 *
 * Behavior:
 * - Executes the 39 list management scenario.
 * - Validates core operations.
 *
 * Expectations:
 * - Scenario completes successfully without errors.
 */

import { getActor, PDS1, SERVICE_URLS } from "../../lib/deno/config.ts";
import { now, ScenarioResult } from "../../lib/deno/runner.ts";
export { ScenarioResult, StepResult, StepStatus } from "../../lib/deno/runner.ts";
export type { ScenarioReport } from "../../lib/deno/runner.ts";
import { XrpcClient, XrpcError } from "../../lib/deno/client.ts";
import { assert } from "../../lib/deno/assertions.ts";
import { timedCall } from "../../lib/deno/runner.ts";

/**
 * Executes the scenario logic.
 * @returns A promise that resolves to the scenario result
 */


function delay(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

/** Retry a predicate function with timeout. Throws if not fulfilled within deadline. */
async function waitFor(
  predicate: () => Promise<boolean>,
  timeoutMs = 15000,
  intervalMs = 500,
): Promise<void> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    if (await predicate()) return;
    await delay(intervalMs);
  }
  if (!(await predicate())) {
    throw new Error(`Condition not met within ${timeoutMs}ms`);
  }
}

export async function run(): Promise<ScenarioResult> {
  const result = new ScenarioResult("List Management");
  result.start();

  const pds = new XrpcClient(PDS1);
  const appview = new XrpcClient(SERVICE_URLS.appview);
  const luna = getActor("luna");
  const marcus = getActor("marcus");

  await timedCall(result, "PDS health check", async () => {
    await pds.waitForHealthy(30);
  });

  if (result.failed > 0) return result;

  for (const char of [luna, marcus]) {
    const session = await pds.accounts.createAccount(char.handle, char.email, char.password).catch(
      () => pds.accounts.createSession(char.handle, char.password),
    );
    if (session) {
      char.did = session.did;
      char.accessJwt = session.accessJwt;
    }
  }

  const listRkey = `curate-list-${Date.now()}`;
  const listRecord = {
    $type: "app.bsky.graph.list",
    purpose: "app.bsky.graph.defs#curatelist",
    name: "Luna's Favorites",
    description: "Accounts Luna finds interesting",
    createdAt: now(),
  };

  const listRef = await timedCall(result, "Create curate list", async () => {
    return await pds.records.createRecord(
      luna.did,
      "app.bsky.graph.list",
      listRecord,
      luna.accessJwt,
      { rkey: listRkey },
    );
  });

  if (listRef) {
    const listUri = listRef.uri;
    const itemRkey = `item-${Date.now()}`;

    await timedCall(result, "Add Marcus to list", async () => {
      return await pds.records.createRecord(
        luna.did,
        "app.bsky.graph.listitem",
        {
          $type: "app.bsky.graph.listitem",
          list: listUri,
          subject: marcus.did,
          createdAt: now(),
        },
        luna.accessJwt,
        { rkey: itemRkey },
      );
    });

    await timedCall(result, "Get lists for Luna", async () => {
      let lastCount = 0;
      await waitFor(async () => {
        try {
          const res = await appview.as(luna).raw.get(
            "app.bsky.graph.getLists",
            { actor: luna.did, limit: 10 },
          );
          lastCount = res.lists?.length || 0;
          return lastCount > 0;
        } catch {
          return false;
        }
      });
      return { lists: lastCount };
    });

    await timedCall(result, "Get list items", async () => {
      let lastCount = 0;
      await waitFor(async () => {
        try {
          const res = await appview.as(luna).raw.get(
            "app.bsky.graph.getList",
            { list: listUri, limit: 10 },
          );
          lastCount = res.items?.length || 0;
          return true; // getList succeeded (list is indexed)
        } catch {
          return false;
        }
      });
      return { items: lastCount };
    });

    await timedCall(result, "Remove Marcus from list", async () => {
      return await pds.records.deleteRecord(
        luna.did,
        "app.bsky.graph.listitem",
        itemRkey,
        luna.accessJwt,
      );
    });
  }

  result.finish();
  return result;
}

if (import.meta.main) {
  run().then((res) => {
    console.log(res.summary());
    Deno.exit(res.ok ? 0 : 1);
  });
}
